### PR TITLE
Python 3.11 on mac/manylinux (with updated cibuildwheel)

### PR DESCRIPTION
**Pull request by ankith26**
Updates cibuildwheel so that we can release the next dev release with 3.11 (rc2) support on mac/manylinux. This is a release candidate, but since it is intended to be ABI compatible with the final release, python encourages library devs to release with these.

~~Also removes manylinux1_i686 and manylinux2010_i686. The rationale is that manylinux1 and manylinux2010 wheels are needed by (probably) less than 5% of our users, and 32bit linux looks like it is almost non-existent these days. Besides, even if we have a user on a 32bit linux and they come with a bug report, we can always recommend them to update pip so that it picks up the manylinux2014_i686 wheel (which we should still keep around for some time IMHO)~~

~~Also does some cleanups and bugfixes (the pull targets were fixed, and manylinux1 was uncommented out) to some manylinux buildconfig code~~
EDIT: the latter have been deferred to future PRs

### CODE REVIEW

File: .github/workflows/build-macos.yml [hunk 1]
```diff
@@ -75,36 +75,37 @@ jobs:
...
-            pyversions: cp39-* pp39-*
+            pyversions: "?p39-*"
           }
 
           - { 
-            name: "arm64 (cp38-cp39-cp310)",
+            name: "arm64 (CPython 3.8 and above)",
             macarch: arm64,
-            pyversions: cp38-* cp39-* cp310-*
+            pyversions: "cp3{8,9,[1-9][0-9]}-*"
```
Comment (Starbuck5) [inline .github/workflows/build-macos.yml:108]: Using globs seems like it makes it less clear what exactly does what.

Comment (ankith26) [reply .github/workflows/build-macos.yml:108]: Well, I think the more compact representation is better. Besides, there's a human readable name just a few lines above each of those globs anyways

Comment (Starbuck5) [reply .github/workflows/build-macos.yml:108]: compact representation? I expected you to talk about forwards compatibility or something. Explicit is better than implicit?

File: buildconfig/manylinux-build/build-wheels.sh [hunk 2]
```diff
@@ -1,17 +1,12 @@
 #!/bin/bash
 set -e -x
 
+# This file is not used by CI anymore. This only exists incase someone wants to
+# generate manylinux wheels locally for testing
```
Comment (Starbuck5) [inline buildconfig/manylinux-build/build-wheels.sh:5]: What's the point of all the changes to this file if it's not being used anymore in CI? Would anyone actually use this locally? Have you tested these changes locally?

File: .github/workflows/build-macos.yml (hunk 1 above)
Comment (ankith26) [reply .github/workflows/build-macos.yml:108]: > I expected you to talk about forwards compatibility or something.

Yeah well that's another thing, but as you know in a year when we have to add 3.12 here a lot more things could change (like having to accommodate a new pypy version, or drop an older one) and we would probably have to move around stuff here a bit anyways.

Also I don't see how this is less clear, but this is subjective anyways. I see this pattern being used in many other projects using cibuildwheel (and also used in cibuildwheel docs) so I assumed people won't have much of a problem with it. Also this is not really an "implicit" thing.

File: buildconfig/manylinux-build/build-wheels.sh (hunk 2 above)
Comment (ankith26) [reply buildconfig/manylinux-build/build-wheels.sh:5]: Can't speak for others, but I do use this file locally, and often with a lot of custom changes ;)
I've also tested these changes

File: .github/workflows/build-macos.yml (hunk 1 above)
Comment (Starbuck5) [reply .github/workflows/build-macos.yml:108]: I'm not going to be a complete stick in the mud about this from the perspective of approving this PR, but this is absolutely an implicit thing.

Explicit: `cp38-* cp39-* cp310-*`
Implicit: `cp3{8,9,[1-9][0-9]}-*`

Comment (ankith26) [reply .github/workflows/build-macos.yml:108]: Well, there's still a `*` at the end of what you call "explicit" so I don't really see how this is an implicit vs explicit thing
Explicit is better than implicit, but practicality beats purity :wink:

Comment (Starbuck5) [reply .github/workflows/build-macos.yml:108]: All I'm saying is that to actually understand the new style, you have to translate it in your head into the old style, and then you can think about it.

File: .github/workflows/build-manylinux.yml [hunk 3]
```diff
@@ -32,24 +32,23 @@ on:
...
-          - { image: manylinux2010, arch: i686, pyversions: cp* pp37-* pp38-* }
+          # no pypy >= 3.9 and cpython >= cp311 on manylinux2010
+          - { image: manylinux2010, arch: x86_64, pyversions: "cp3{[6-9],10}-* pp3{7,8}-*" }
 
-          # all cpython and pypy versions should support this
-          - { image: manylinux2014, arch: x86_64, pyversions: cp* pp* }
-          - { image: manylinux2014, arch: i686, pyversions: cp* pp* }
+          # all cpython and pypy versions should support this for now
+          - { image: manylinux2014, arch: x86_64, pyversions: "*" }
+          - { image: manylinux2014, arch: i686, pyversions: "*" }
```
Comment (Starbuck5) [inline .github/workflows/build-manylinux.yml:51]: ALL?

You mean all valid versions from the perspective of cibuildwheel? What does that mean?

File: .github/workflows/build-macos.yml [hunk 4]
```diff
@@ -75,36 +75,37 @@ jobs:
       fail-fast: false  # if a particular matrix build fails, don't skip the rest
       matrix:
         # Split job into 5 matrix builds, because GH actions provides 5 concurrent
-        # builds on macOS
+        # builds on macOS. This needs to be manually kept updated so that each
+        # of these builds take roughly the same time
         include:
           - { 
-            name: "x86_64 (cp36-cp310)",
+            name: "x86_64 (CPython 3.6 3.10 and above)",
```
Comment (Starbuck5) [inline .github/workflows/build-macos.yml:82]: Isn't this CPython 3.6 and above?

Right now it sounds like: 3.6, 3.10, and 3.11+

Comment (Starbuck5) [reply .github/workflows/build-macos.yml:82]: Ah geez this is 3.6 or 3.[double digit number].

That's hard to parse for me.

**Comment by Starbuck5** [comment]
Can you explain what gettext is and why it's changing?

File: .circleci/config.yml [hunk 5]
```diff
@@ -14,7 +14,6 @@ jobs:
     environment:
       # these environment variables will be passed to the docker container
       - CIBW_ENVIRONMENT: PIP_CONFIG_FILE=buildconfig/pip_config.ini PORTMIDI_INC_PORTTIME=1 SDL_VIDEODRIVER=dummy SDL_AUDIODRIVER=disk
-      - CIBW_BUILD: cp* pp*
```
Comment (Starbuck5) [inline .circleci/config.yml:17]: I don't understand why this change is here.

File: .github/workflows/build-manylinux.yml (hunk 3 above)
Comment (ankith26) [reply .github/workflows/build-manylinux.yml:51]: All versions the specific cibuildwheel version supports

File: .github/workflows/build-macos.yml (hunk 4 above)
Comment (ankith26) [reply .github/workflows/build-macos.yml:82]: It is what it sounds

File: .circleci/config.yml (hunk 5 above)
Comment (ankith26) [reply .circleci/config.yml:17]: That line was basically redundant because cibuildwheel default is to just do that

**Comment by ankith26** [comment]
Basically, `gettext` is an low-level library for handling text things used by glib, which in turn is used by many of our dependencies (most notably by fluidsynth, which is in turn used by SDL_mixer)

What changed when I bumped cibuildwheel, is that the python it uses internally started depending on `/usr/local/opt/gettext`. But that was being removed as part of our mac dep builds because we install our own gettext on arm64 (in `/usr/local/gettext`, which system python can't use). So cibuildwheel started failing on startup, so I removed the line that was purging `/usr/local/opt/gettext`. Now we have both gettexts peacefully coexisting, the system python picks the one in opt (intel), and our builds pick what we install (arm64)

File: .github/workflows/build-manylinux.yml (hunk 3 above)
Comment (Starbuck5) [reply .github/workflows/build-manylinux.yml:51]: So right now that happens to line up with what we support 3.6, but it won't always? I know that we have a pinned dependency version, but I'm still thinking about that.

Comment (ankith26) [reply .github/workflows/build-manylinux.yml:51]: Till someone goes ahead and bumps the cibuildwheel version, nothing is going to change here. Even when cibuildwheel drops 3.6 in a version, we won't be affected till we bump it on our side (which is why pinned dependencies are ideal)

Also in this particular instance `cp* pp*` is basically equivalent to `*` but IMHO more readable as it is immediately clear that this tells cibuildwheel to use everything it's got

Comment (ankith26) [reply .github/workflows/build-manylinux.yml:51]: Till someone goes ahead and bumps the cibuildwheel version, nothing is going to change here. Even when cibuildwheel drops 3.6 in a version, we won't be affected till we bump it on our side (which is why pinned dependencies are ideal)

Also in this particular instance `cp* pp*` is basically equivalent to `*` but IMHO more readable as it is immediately clear that this tells cibuildwheel to use everything it's got

Comment (Starbuck5) [reply .github/workflows/build-manylinux.yml:51]: Leave a comment though just for future people?

**Review by Starbuck5 — APPROVED** [review]
This PR was hard to review, because it's stuff I'm not familiar with, and because there are a lot of changes in this PR besides just adding 3.11 support. I get that everything is in the same vein, but it would've been easier perhaps as smaller PRs that just did the 1 thing? Idk.

Anyways, I asked a bunch of questions and got a bunch of answers, so I'll go ahead and approve. Thanks for working on this Ankith.

**Comment by illume** [comment]
I'd prefer if we keep this for after 2.1.3. Cool?

**Comment by ankith26** [comment]
The old 32bit manylinux wheels drop, I definitely don't mind postponing it because it's more of a cleanup thing anyways. But I would like 2.1.3 to have python 3.11 support on all platforms (and we have it on windows already, so this PR is meant to be a follow-up on mac/manylinux)

I could open a separate, minimal PR for that if that's okay with you

**Comment by ankith26** [comment]
(this is the relevant windows PR #3486)

**Comment by illume** [comment]
Yeah, cool thanks. One without dropping would be wonderful.

**Comment by ankith26** [comment]
Now this PR is doing only the 3.11 bump

I also removed some of the minor docker file changes I did here, but the issue was that some manylinux1 stuff was commented out (which affected dev6, manylinux1 is still lagging behind in there)

**Review by illume — APPROVED** [review]
Awesome sauce. Thanks :)
